panic when loading volumes concurrently

**Describe the bug**
Loading same volume more than once leads to SIGSEGV

**System Setup**
- List the command line to start "weed master", "weed volume", "weed filer", "weed s3", "weed mount".
- OS version
- output of `weed version`
- if using filer, show the content of `filer.toml`

1. start server: `./weed server`
2. upload some random files
3. restart with "-volume.index leveldb"
 `./weed server -volume.index leveldb`

4. crash log

> I1216 18:24:45 91019 master_grpc_server.go:250] + client master@10.170.160.252:56816
I1216 18:24:45 91019 volume_loading.go:128] loading leveldb /var/folders/1x/lbz9zhkd1_n0vd2zlmsl_0n40000gn/T/2.ldb
I1216 18:24:45 91019 volume_loading.go:128] loading leveldb /var/folders/1x/lbz9zhkd1_n0vd2zlmsl_0n40000gn/T/4.ldb
I1216 18:24:45 91019 volume_loading.go:128] loading leveldb /var/folders/1x/lbz9zhkd1_n0vd2zlmsl_0n40000gn/T/3.ldb
I1216 18:24:45 91019 volume_loading.go:128] loading leveldb /var/folders/1x/lbz9zhkd1_n0vd2zlmsl_0n40000gn/T/2.ldb
I1216 18:24:45 91019 volume_loading.go:128] loading leveldb /var/folders/1x/lbz9zhkd1_n0vd2zlmsl_0n40000gn/T/3.ldb
I1216 18:24:45 91019 volume_loading.go:128] loading leveldb /var/folders/1x/lbz9zhkd1_n0vd2zlmsl_0n40000gn/T/1.ldb
I1216 18:24:45 91019 volume_loading.go:128] loading leveldb /var/folders/1x/lbz9zhkd1_n0vd2zlmsl_0n40000gn/T/1.ldb
I1216 18:24:45 91019 volume_loading.go:128] loading leveldb /var/folders/1x/lbz9zhkd1_n0vd2zlmsl_0n40000gn/T/4.ldb
I1216 18:24:45 91019 volume_loading.go:128] loading leveldb /var/folders/1x/lbz9zhkd1_n0vd2zlmsl_0n40000gn/T/5.ldb
I1216 18:24:45 91019 volume_loading.go:128] loading leveldb /var/folders/1x/lbz9zhkd1_n0vd2zlmsl_0n40000gn/T/5.ldb
I1216 18:24:45 91019 volume_loading.go:135] loading leveldb /var/folders/1x/lbz9zhkd1_n0vd2zlmsl_0n40000gn/T/4.ldb error: resource temporarily unavailable
I1216 18:24:45 91019 volume_loading.go:135] loading leveldb /var/folders/1x/lbz9zhkd1_n0vd2zlmsl_0n40000gn/T/2.ldb error: resource temporarily unavailable
I1216 18:24:45 91019 volume_loading.go:135] loading leveldb /var/folders/1x/lbz9zhkd1_n0vd2zlmsl_0n40000gn/T/3.ldb error: resource temporarily unavailable
I1216 18:24:45 91019 volume_loading.go:135] loading leveldb /var/folders/1x/lbz9zhkd1_n0vd2zlmsl_0n40000gn/T/1.ldb error: resource temporarily unavailable
I1216 18:24:45 91019 volume_loading.go:135] loading leveldb /var/folders/1x/lbz9zhkd1_n0vd2zlmsl_0n40000gn/T/5.ldb error: resource temporarily unavailable
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x188 pc=0x49e80be]

**Expected behavior**
not crash

--- a/weed/storage/disk_location.go
+++ b/weed/storage/disk_location.go
@@ -104,19 +104,19 @@ func (l *DiskLocation) loadExistingVolume(fileInfo os.FileInfo, needleMapKind Ne
 		l.volumesLock.Unlock()
 		glog.V(1).Infof("loaded volume, %v", vid)
 		return true
+	} else {
+		l.volumes[vid] = nil
+		l.volumesLock.Unlock()
 	}
 
 	// load the volume
 	v, e := NewVolume(l.Directory, l.IdxDirectory, collection, vid, needleMapKind, nil, nil, 0, 0)
 	if e != nil {
-		l.volumesLock.Unlock()
 		glog.V(0).Infof("new volume %s error %s", volumeName, e)
 		return false
 	}
-
-	l.volumes[vid] = v
-	v.location = l
-	l.volumesLock.Unlock()
+	
+	l.SetVolume(vid, v)
 
 	size, _, _ := v.FileStat()
 	glog.V(0).Infof("data file %s, replicaPlacement=%s v=%d size=%d ttl=%s",

--- a/weed/storage/disk_location.go
+++ b/weed/storage/disk_location.go
@@ -99,10 +99,9 @@ func (l *DiskLocation) loadExistingVolume(fileInfo os.FileInfo, needleMapKind Ne
 	}
 
 	// avoid loading one volume more than once
-	l.volumesLock.RLock()
-	_, found := l.volumes[vid]
-	l.volumesLock.RUnlock()
-	if found {
+	l.volumesLock.Lock()
+	if _, found := l.volumes[vid]; found {
+		l.volumesLock.Unlock()
 		glog.V(1).Infof("loaded volume, %v", vid)
 		return true
 	}
@@ -110,11 +109,14 @@ func (l *DiskLocation) loadExistingVolume(fileInfo os.FileInfo, needleMapKind Ne
 	// load the volume
 	v, e := NewVolume(l.Directory, l.IdxDirectory, collection, vid, needleMapKind, nil, nil, 0, 0)
 	if e != nil {
+		l.volumesLock.Unlock()
 		glog.V(0).Infof("new volume %s error %s", volumeName, e)
 		return false
 	}
 
-	l.SetVolume(vid, v)
+	l.volumes[vid] = v
+	v.location = l
+	l.volumesLock.Unlock()
 
 	size, _, _ := v.FileStat()
 	glog.V(0).Infof("data file %s, replicaPlacement=%s v=%d size=%d ttl=%s",

